modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticDruckerPrager.C

```diff
@@ -0,0 +1,275 @@
+  if (_mc_cohesion.value(0) < 0)
+    mooseError("TensorMechanicsPlasticDruckerPrager: MC cohesion should not be negative");
+
+  initializeAandB(0.0, _aaa, _bbb);
+  initializeB(0.0, dilation, _bbb_flow);
+}
+
+
+Real
+TensorMechanicsPlasticDruckerPrager::yieldFunction(const RankTwoTensor & stress, Real intnl) const
+{
+  Real aaa;
+  Real bbb;
+  AandB(intnl, aaa, bbb);
+  return std::sqrt(stress.secondInvariant()) + stress.trace() * bbb - aaa;
+}
+
+RankTwoTensor
+TensorMechanicsPlasticDruckerPrager::df_dsig(const RankTwoTensor & stress, Real bbb) const
+{
+  return 0.5 * stress.dsecondInvariant()/std::sqrt(stress.secondInvariant()) + stress.dtrace() * bbb;
+}
+
+RankTwoTensor
+TensorMechanicsPlasticDruckerPrager::dyieldFunction_dstress(const RankTwoTensor & stress, Real intnl) const
+{
+  Real bbb;
+  Bonly(intnl, friction, bbb);
+  return df_dsig(stress, bbb);
+}
+
+
+Real
+TensorMechanicsPlasticDruckerPrager::dyieldFunction_dintnl(const RankTwoTensor & stress, Real intnl) const
+{
+  Real daaa;
+  Real dbbb;
+  dAandB(intnl, daaa, dbbb);
+  return stress.trace() * dbbb - daaa;
+}
+
+RankTwoTensor
+TensorMechanicsPlasticDruckerPrager::flowPotential(const RankTwoTensor & stress, Real intnl) const
+{
+  Real bbb_flow;
+  Bonly(intnl, dilation, bbb_flow);
+  return df_dsig(stress, bbb_flow);
+}
+
+RankFourTensor
+TensorMechanicsPlasticDruckerPrager::dflowPotential_dstress(const RankTwoTensor & stress, Real /*intnl*/) const
+{
+  RankFourTensor dr_dstress;
+  dr_dstress = 0.5 * stress.d2secondInvariant() / std::sqrt(stress.secondInvariant());
+  dr_dstress += -0.5 * 0.5 * stress.dsecondInvariant().outerProduct(stress.dsecondInvariant()) / std::pow(stress.secondInvariant(), 1.5);
+  return dr_dstress;
+}
+
+RankTwoTensor
+TensorMechanicsPlasticDruckerPrager::dflowPotential_dintnl(const RankTwoTensor & stress, Real intnl) const
+{
+  Real dbbb;
+  dBonly(intnl, dilation, dbbb);
+  return stress.dtrace() * dbbb;
+}
+
+
+std::string
+TensorMechanicsPlasticDruckerPrager::modelName() const
+{
+  return "DruckerPrager";
+}
+
+
+void
+TensorMechanicsPlasticDruckerPrager::AandB(Real intnl, Real & aaa, Real & bbb) const
+{
+  if (_zero_cohesion_hardening && _zero_phi_hardening)
+  {
+    aaa = _aaa;
+    bbb = _bbb;
+    return;
+  }
+  initializeAandB(intnl, aaa, bbb);
+}
+
+void
+TensorMechanicsPlasticDruckerPrager::Bonly(Real intnl, int fd, Real & bbb) const
+{
+  if (_zero_phi_hardening && (fd == friction))
+  {
+    bbb = _bbb;
+    return;
+  }
+  if (_zero_psi_hardening && (fd == dilation))
+  {
+    bbb = _bbb_flow;
+    return;
+  }
+  initializeB(intnl, fd, bbb);
+}
+
+
+void
+TensorMechanicsPlasticDruckerPrager::dBonly(Real intnl, int fd, Real & dbbb) const
+{
+  if (_zero_phi_hardening && (fd == friction))
+  {
+    dbbb = 0;
+    return;
+  }
+  if (_zero_psi_hardening && (fd == dilation))
+  {
+    dbbb = 0;
+    return;
+  }
+  const Real sin = (fd == friction) ? std::sin(_mc_phi.value(intnl)) : std::sin(_mc_psi.value(intnl));
```

Comment:
calling this symbol `sin` might be a little dangerous (if somewhere along the include chain `math.h` is ever included)